Support wildcards in probe path

acj opened the pull request:
Fixes #688. One case that remains tricky is `usdt:/my/app/*` because `path` includes leading and trailing colons in the lexer. It works if you provide a probe name, though. Maybe this is fine for now.

I'm not sure whether all the tests are needed or wanted, but they were helpful for testing, so I left them in.

ajor commented:
Looking at the current parser tests, I see there's a one for "wildcard_attach_points" at the moment, so I think it'd be better to put your new ones in a new "wildcard_path" test case.

Would also be good to add some semantic analyser tests to catch the bug I mentioned in the last comment. Would need to mock out `resolve_binary_path()` (or similar) to avoid depending on a user's filesystem layout.

acj commented:
Okay, sounds good. I was able to attach to copies of the usdt_test binary in testprogs with this change, but I want to verify that my test was actually matching more than one binary. The same bug may exist for all providers.

Thanks for the feedback.

acj commented:
I cleaned up the tests and fixed a couple of issues that turned up in my usage testing.

@ajor For the scenario you mentioned, this is the output:
```
# ls /tmp
a.out      b.out

# bpftrace -e 'uprobe:/tmp/*.out:main { printf("hello %s\n", comm) }'
path '/tmp/*.out' must refer to a unique binary but matched 2
```

Listing probes will now display the expanded path, i.e. without wildcards:

```
# bpftrace -l 'uprobe:/tmp/a.*:main'
uprobe:/tmp/a.out:main
uprobe:/tmp/a.out:main
```

danobi commented on src/ast/semantic_analyser.cpp:
```diff
@@ -996,21 +1000,23 @@ void SemanticAnalyser::visit(AttachPoint &ap)
+    if (ap.target != "") {
+      auto paths = resolve_binary_path(ap.target);
+      if (paths.size() > 1)
+	err_ << "path '" << ap.target << "' must refer to a unique binary but matched " << paths.size() << std::endl;
```
Indent please

acj commented:
@danobi Fixed and rebased. Nice catch on the indentation.

The test cases that I removed in the final commit apparently only work when `/bin/sh` is bash. CI must have been upgraded to a newer Linux that uses a different shell (dash?) after I opened this. Using dash as my shell locally, I see that its behavior for `command -v ...` is different from bash's in that dash will only return a single result even if there are several matches. Bash's version will return all matches.

I'd like to work on removing our dependency on `command` in a followup PR if that sounds good to you.

danobi commented:
@acj Ah got it. Yeah, that would be great if you could remove `command -v` shellout. Seems like that was a bad idea after all.

Thanks for the patch!
